feat(go/adbc/driver/flightsql): Native Golang ADBC Flight SQL driver

zeroshade opened the pull request:
Not every feature is implemented yet, but the vast majority is here along with the basic framework of a validation testing suite for drivers.

lidavidm commented on go/adbc/driver/flightsql/flightsql_adbc.go:
```diff
@@ -0,0 +1,450 @@
+// You can then open a flightsql connection with the database/sql
+// standard package by using:
+//
+//     db, err := sql.Open("flightsql", "adbc.url=<flight sql db url>")
```
Thought the option was just 'uri'?

lidavidm reviewed (commented):
Generally seems good to me as a start.

zeroshade commented:
Updated with further implementations, all that is left is to write tests. I should have tests written out tomorrow and updated. Then i'll mark this as ready for review.

lidavidm commented on go/adbc/driver/flightsql/record_reader.go:
```diff
@@ -0,0 +1,132 @@
+	cancelFn context.CancelFunc
+}
+
+// kicks off a goroutine for each endpoint and returns a reader which
```
I am jealous how easily you can do this...I'm dreading the C++ version

lidavidm commented on go/adbc/driver/flightsql/flightsql_adbc_test.go:
```diff
@@ -0,0 +1,201 @@
+
+func (s srvQuery) GetQuery() string { return string(s) }
+
+func writeTo(arr arrow.Array, idx int, w io.Writer) {
```
Huh, this isn't built into the Arrow library itself?

zeroshade replied:
the Arrow library when printing an array uses the Golang slice syntax, ie: no commas. So if I want commas in between the values, I need to do it myself.

lidavidm commented on go/adbc/driver/flightsql/flightsql_adbc.go:
```diff
@@ -0,0 +1,577 @@
+
+const (
+	OptionSSLInsecure = "adbc.flight.sql.client_option.tls_skip_verify"
+	OptionSSLCertFile = "adbc.flight.sql.client_option.tls_root_certs"
```
I wonder if we want to align between the C++/Java/Go Flight SQL drivers?

lidavidm commented on go/adbc/driver/flightsql/flightsql_adbc.go:
```diff
@@ -0,0 +1,577 @@
+	OptionSSLInsecure = "adbc.flight.sql.client_option.tls_skip_verify"
+	OptionSSLCertFile = "adbc.flight.sql.client_option.tls_root_certs"
+
+	infoDriverName = "ADBC Flight SQL Driver - Go"
```
Heh, the C++/Java ones should also probably include the language then.

zeroshade replied:
Particularly if we're going to do the whole packaging the Go one into a shared library, it's good to include the language to make it easier to debug.

lidavidm commented on .pre-commit-config.yaml:
```diff
@@ -52,8 +52,8 @@ repos:
     - id: golangci-lint
-      entry: bash -c 'cd go/adbc && golangci-lint run --fix'
-      args: [--fix]
+      entry: bash -c 'cd go/adbc && golangci-lint run --fix --timeout 2m'
```
we're hitting the timeout??

zeroshade replied:
The default timeout is 1m, and i can't reproduce hitting the timeout locally. The only thing I can think of is that behind the scenes it's grabbing the dependencies or something and *that's* taking longer? I don't know. It's really strange and I haven't been able to reproduce it anywhere but in the GHA runner....

lidavidm replied:
Well, I'd believe that the runner is just slower than our local machines so that's OK.

lidavidm reviewed (commented):
LGTM in general. Left some very minor comments.

lidavidm commented:
What do you think about just moving the Go constant into the driver itself for now, rebasing out the format commit, and merging?

lidavidm reviewed (approved):
LGTM - this is exciting!

zeroshade commented:
@lidavidm Sure I'll set that up and push it without the format commit

jacobmarble commented:
Thank you, @zeroshade
